src/autocomplete/shared/extract-statement-positions-from-query.ts

```diff
@@ -0,0 +1,148 @@
+import * as c3 from 'antlr4-c3';
+import type {Lexer as LexerType, Parser as ParserType, Token, TokenStream} from 'antlr4ng';
+
+import {GetParseTree, LexerConstructor, ParserConstructor} from './autocomplete-types';
+import {createParser} from './query';
+import {SqlErrorListener} from './sql-error-listener';
+
+export interface StatementPosition {
+    startIndex: number;
+    endIndex: number;
+}
+
+export enum StatementExtractionStrategy {
+    Autocomplete = 'autocomplete',
+    Tokens = 'tokens',
+}
+
+export interface ExtractStatementPositionsResult {
+    statementPositions: StatementPosition[];
+    strategy: StatementExtractionStrategy;
+}
+
+export function extractStatementPositionsFromQuery<L extends LexerType, P extends ParserType>(
+    query: string,
+    Lexer: LexerConstructor<L>,
+    Parser: ParserConstructor<P>,
+    whitespaceToken: number,
+    emptySpaceTokens: number[],
+    endStatementToken: number,
+    statementRule: number,
+    getParseTree: GetParseTree<P>,
+): ExtractStatementPositionsResult {
+    const parser = createParser(Lexer, Parser, query);
+    const {tokenStream} = parser;
+    const errorListener = new SqlErrorListener(whitespaceToken);
+
+    parser.removeErrorListeners();
+    parser.addErrorListener(errorListener);
+    getParseTree(parser);
+
+    const autocompleteStatementPositions = extractStatementsUsingAutocomplete(
+        parser,
+        tokenStream,
+        statementRule,
+    );
+    if (autocompleteStatementPositions.length) {
+        return {
+            statementPositions: autocompleteStatementPositions,
+            strategy: StatementExtractionStrategy.Autocomplete,
+        };
+    }
+
+    return extractStatementsUsingTokens(tokenStream, emptySpaceTokens, endStatementToken);
+}
+
+export function extractStatementsUsingTokens(
+    tokenStream: TokenStream,
+    emptySpaceTokens: number[],
+    endStatementToken: number,
+): ExtractStatementPositionsResult {
+    let statementStartIndex = 0;
+    let processingNewStatement = false;
+
+    const statementPositions: StatementPosition[] = [];
+
+    for (let index = 0; index < tokenStream.size - 1; index++) {
```

Comment:
nit: I think better to make variable `lastTokenPosition` which will equal to `tokenStream.size - 2`, because actually we have duplicates here

how it may look
```ts
for (let index = 0; index <= lastTokenPosition; index++) {
```
85 row
```ts
const isLastToken = index === lastTokenPosition;
```